File: wallendar-app/app/src/main/java/com/td/wallendar/home/balances/BalanceAdapter.java
package com.td.wallendar.home.balances;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.td.wallendar.R;
import com.td.wallendar.models.Debt;

import java.util.ArrayList;
import java.util.List;

public class BalanceAdapter extends RecyclerView.Adapter<BalanceViewHolder> {

    private final List<Debt> dataset;
    private final long loggedUserId;
    private OnBalanceSettleUpClickedListener onBalanceSettleUpClickedListener;
    private OnRemindButtonClickedListener onRemindButtonClickedListener;

    public BalanceAdapter(final long loggedUserId) {
        this.dataset = new ArrayList<>();
        this.loggedUserId = loggedUserId;
    }

    public void setOnBalanceSettleUpClickedListener(OnBalanceSettleUpClickedListener listener, OnRemindButtonClickedListener onRemindButtonClickedListener) {
        this.onBalanceSettleUpClickedListener = listener;
        this.onRemindButtonClickedListener = onRemindButtonClickedListener;
    }

    @NonNull
    @Override
    public BalanceViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_balances, parent, false);
        return new BalanceViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull BalanceViewHolder holder, int position) {
        holder.bind(dataset.get(position), loggedUserId);
        holder.setOnBalanceSettleUpClickedListener(onBalanceSettleUpClickedListener);
        holder.setOnRemindButtonClickedListener(onRemindButtonClickedListener);
    }

    @Override
    public int getItemCount() {
        return dataset == null ? 0 : dataset.size();
    }

    @SuppressLint("NotifyDataSetChanged")
    public void setDataset(final List<Debt> newDataset) {
        this.dataset.clear();
        if (newDataset != null) {
            this.dataset.addAll(newDataset);
        }
        notifyDataSetChanged();
    }

    public void removeFromDataset(Debt debt) {
        int position = this.dataset.indexOf(debt);
        this.dataset.remove(position);
        notifyItemRemoved(position);
    }
}

File: wallendar-app/app/src/main/java/com/td/wallendar/home/balances/BalanceAdapter.kt
package com.td.wallendar.home.balances

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.td.wallendar.R
import com.td.wallendar.models.Debt
import java.util.*

class BalanceAdapter(loggedUserId: Long) : RecyclerView.Adapter<BalanceViewHolder?>() {
    private val dataset: MutableList<Debt>
    private val loggedUserId: Long
    private var onBalanceSettleUpClickedListener: OnBalanceSettleUpClickedListener? = null
    private var onRemindButtonClickedListener: OnRemindButtonClickedListener? = null
    private var onDetailDebtClickedListener: OnDetailDebtClickedListener? = null

    fun setOnBalanceSettleUpClickedListener(
            listener: OnBalanceSettleUpClickedListener?,
            onRemindButtonClickedListener: OnRemindButtonClickedListener?,
            onDetailDebtClickedListener: OnDetailDebtClickedListener?,
    ) {
        this.onBalanceSettleUpClickedListener = listener
        this.onRemindButtonClickedListener = onRemindButtonClickedListener
        this.onDetailDebtClickedListener = onDetailDebtClickedListener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BalanceViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.row_balances, parent, false)
        return BalanceViewHolder(view)
    }

    override fun onBindViewHolder(holder: BalanceViewHolder, position: Int) {
        holder.bind(dataset[position], loggedUserId)
        holder.setOnBalanceSettleUpClickedListener(onBalanceSettleUpClickedListener)
        holder.setOnRemindButtonClickedListener(onRemindButtonClickedListener)
        holder.setOnDetailDebtClickedListener(onDetailDebtClickedListener)
    }

    override fun getItemCount(): Int {
        return dataset.size
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setDataset(newDataset: MutableList<Debt>) {
        dataset.clear()
        dataset.addAll(newDataset)
        notifyDataSetChanged()
    }

    fun removeFromDataset(debt: Debt?) {
        val position = dataset.indexOf(debt)
        dataset.removeAt(position)
        notifyItemRemoved(position)
    }

    init {
        dataset = ArrayList()
        this.loggedUserId = loggedUserId
    }
}